From 7adfbdbc8bf56aaa929a68caf1110aa8515d742c Mon Sep 17 00:00:00 2001
From: Martin Blumenstingl <martin.blumenstingl@googlemail.com>
Date: Sat, 25 Dec 2021 04:18:30 +0100
Subject: [PATCH 10/98] net: stmmac: dwmac-meson: Initialize all known
 PREG_ETHERNET_ADDR0 bits

Initialize all known PREG_ETHERNET_ADDR0 register bits to be less
dependent on the bootloader to set them up correctly.

Signed-off-by: Martin Blumenstingl <martin.blumenstingl@googlemail.com>

[dts: rebased onto 6.6.144 for 6.6-lts - dropped meson6_dwmac_set_phy_intf_sel
 and plat_dat->set_phy_intf_sel: that stmmac core callback/PHY_INTF_SEL_*
 enum doesn't exist at 6.6.144 (same later stmmac refactor as set_clk_tx_rate,
 see 009). This board already works with the bootloader's PHY_INTF_SEL default
 (RMII), so only the endianness-bit hardening is kept.]
---
diff --git a/drivers/net/ethernet/stmicro/stmmac/dwmac-meson.c b/drivers/net/ethernet/stmicro/stmmac/dwmac-meson.c
index 6a7207721..8cc404066 100644
--- a/drivers/net/ethernet/stmicro/stmmac/dwmac-meson.c
+++ b/drivers/net/ethernet/stmicro/stmmac/dwmac-meson.c
@@ -5,6 +5,7 @@
  * Copyright (C) 2014 Beniamino Galvani <b.galvani@gmail.com>
  */
 
+#include <linux/bitfield.h>
 #include <linux/clk.h>
 #include <linux/device.h>
 #include <linux/ethtool.h>
@@ -16,10 +17,18 @@
 
 #include "stmmac_platform.h"
 
-#define PREG_ETHERNET_ADDR0_DIV_EN	BIT(0)
+#define PREG_ETHERNET_ADDR0_DIV_EN			BIT(0)
 
 /* divides the input clock by 20 (= 0x0) or 2 (= 0x1) */
-#define PREG_ETHERNET_ADDR0_SPEED_100	BIT(1)
+#define PREG_ETHERNET_ADDR0_SPEED_100			BIT(1)
+
+/* 0x0 = little, 0x1 = big */
+#define PREG_ETHERNET_ADDR0_DATA_ENDIANNESS		BIT(2)
+
+/* 0x0 = same order, 0x1: unknown */
+#define PREG_ETHERNET_ADDR0_DESC_ENDIANNESS		BIT(3)
+
+#define PREG_ETHERNET_ADDR0_PHY_INTF_SEL		GENMASK(6, 4)
 
 struct meson_dwmac {
 	struct device	*dev;
@@ -50,6 +59,7 @@ static int meson6_dwmac_init(struct platform_device *pdev, void *priv)
 {
 	struct meson_dwmac *dwmac = priv;
 	int ret;
+	u32 val;
 
 	ret = clk_set_rate(dwmac->clk, 50 * 1000 * 1000);
 	if (ret)
@@ -59,7 +69,11 @@ static int meson6_dwmac_init(struct platform_device *pdev, void *priv)
 	if (ret)
 		return ret;
 
-	writel(readl(dwmac->reg) | PREG_ETHERNET_ADDR0_DIV_EN, dwmac->reg);
+	val = readl(dwmac->reg);
+	val &= ~PREG_ETHERNET_ADDR0_DATA_ENDIANNESS;
+	val &= ~PREG_ETHERNET_ADDR0_DESC_ENDIANNESS;
+	val |= PREG_ETHERNET_ADDR0_DIV_EN;
+	writel(val, dwmac->reg);
 
 	return 0;
 }
